albatross/core/concatenate.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 2018 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_CORE_CONCATENATE_H
+#define ALBATROSS_CORE_CONCATENATE_H
+
+#include "core/dataset.h"
+#include "map_utils.h"
+
+namespace albatross {
+
+template <typename FeatureType>
+RegressionDataset<FeatureType> concatenate_datasets(
+    const std::vector<RegressionDataset<FeatureType>> &datasets) {
+  std::size_t n = 0;
+  for (const auto &d : datasets) {
+    n += d.features.size();
+  }
+
+  Eigen::VectorXd target_means(static_cast<Eigen::Index>(n));
+  Eigen::VectorXd target_variance(static_cast<Eigen::Index>(n));
+  std::vector<FeatureType> features(n);
+
+  Metadata metadata;
+  bool target_has_covariance = datasets[0].targets.has_covariance();
+
+  std::size_t i = 0;
+  for (const auto &d : datasets) {
+    assert(target_has_covariance == d.targets.has_covariance());
+    for (std::size_t j = 0; j < d.features.size(); ++j) {
+      features[i] = d.features[j];
+      Eigen::Index ei = static_cast<Eigen::Index>(i);
+      Eigen::Index ej = static_cast<Eigen::Index>(j);
+      target_means[ei] = d.targets.mean[ej];
+      target_variance[ei] = d.targets.get_diagonal(ej);
+      ++i;
+    }
+    metadata = map_join(metadata, d.metadata);
```

Comment:
What happens if two metadatas have the same field, it keeps the latter?
There's no assurance that the metadata for a dataset will apply to the concatenation. This doesn't interfere with anything currently, but it may introduce user error down the line.

Reply:
Yeah exactly, currently if all the metadata is not the same it'll overwrite with the last.  I think if we're explicit about that in a comment then it's OK.  My current philosophy on metadata is that it's around as a helper, but shouldn't really be used or relied on programmatically.  We had a similar approach when designing the merge algorithm for attributes in [`xarray.Dataset`](http://xarray.pydata.org/en/stable/generated/xarray.Dataset.merge.html).  Similar to how that ended up we could eventually provide a strategy enum for concatenation to make the user need to explicitly say something like:
```
enum {
    INTERSECTION_OF_METADATA,
    UNION_OF_METADATA,
    IGNORE_METADATA
} concatenation_strategy_t

...

concatenate_datasets(datasets, INTERSECTION_OF_METADATA)
```
That said, If we're worried about confusion I'm fine just having it drop the metadata for now.